document/JS.bs

```diff
@@ -0,0 +1,921 @@
+
+Note: WebAssembly JS API declaration file for TypeScript can be found [here](https://github.com/01alchemist/webassembly-types/blob/master/webassembly.d.ts) which enable autocompletion and make TypeScript compiler happy.
+
+<h2 id="sample">Sample API usage</h2>
+
+<p><em>This section is non-normative.</em></p>
+
+Given `demo.wat` (encoded to `demo.wasm`):
+
+```lisp
+(module
+    (import "js" "import1" (func $i1))
+    (import "js" "import2" (func $i2))
+    (func $main (call $i1))
+    (start $main)
+    (func (export "f") (call $i2))
+)
+```
+
+and the following JavaScript, run in a browser:
+
+```javascript
+var importObj = {js: {
+    import1: () => console.log("hello,"),
+    import2: () => console.log("world!")
+}};
+fetch('demo.wasm').then(response =>
+    response.arrayBuffer()
+).then(buffer =>
+    WebAssembly.instantiate(buffer, importObj)
+).then(({module, instance}) =>
+    instance.exports.f()
+);
+```
+
+<h2 id="webassembly-namespace">The WebAssembly Namespace</h2>
+
+<pre class="idl">
+typedef (ExportedFunction or double or
+         Memory or Table)
+    WebAssemblyRuntimeValue;
+
+typedef record&lt;DOMString, WebAssemblyRuntimeValue&gt; InstanceExportsMap;
+
+typedef record&lt;DOMString, InstanceExportsMap&gt; InstanceImportsMap;
+
+dictionary WebAssemblyInstantiatedSource {
+   required Module module;
+   required Instance instance;
+};
+
+[Exposed=(Window,Worker,Worklet)]
+namespace WebAssembly {
+  boolean validate(BufferSource bytes);
+  Promise&lt;Module> compile(BufferSource bytes);
+
+  Promise&lt;WebAssemblyInstantiatedSource> instantiate(
+      BufferSource bytes, optional InstanceImportsMap importObject);
+
+  Promise&lt;Instance> instantiate(
+      Module moduleObject, optional InstanceImportsMap importObject);
+};
+</pre>
+
+<!--
+Should we include notes describing what the functions do, as the HTML spec does? It could look like this:
+
+Note:
+  WebAssembly.validate(|bytes|) synchronously validates bytes of WebAssembly, returning true if the validation was successful.
+  WebAssembly.compile(|bytes|) asynchronously validates and complies bytes of WebAssembly into a Module.
+  WebAssembly.instantiate(|bytes|, |importObject|) asynchronously compiles and instantiates a WebAssembly module from bytes of source.
+  The WebAssembly.instantiate(|moduleObject|, |importObject|) asynchronously instantiates a compiled module.
+-->
+
+<div algorithm>
+  To <dfn>compile a WebAssembly module</dfn> from a {{BufferSource}} |bytes|, perform the following steps:
+    1. Let |module| be [=decode_module=](|bytes|). If |module| is [=error=], return [=error=].
+    1. If [=validate_module=](|module|) is [=error=], return [=error=].
+    1. For each <code>name</code> production in |bytes|, including in a custom <code>name</code> section if present, with the subsequence identified as |nameBytes|:
+        1. If running the [=UTF-8 decoder=] algorithm on |nameBytes| leads to an error, return [=error=].
```

Comment:
I think you're not supposed to use this directly. It seems like you want https://encoding.spec.whatwg.org/#utf-8-decode-without-bom-or-fail instead. The naming, at least, is much more explicit.

This happens multiple times throughout.

Reply:
Yeah, specifications should only use the provided hooks and ideally only those with UTF-8 in the name. Interesting that wasm also wants the hard failure mode, just like WebSocket. Most of the platform uses never-failing UTF-8 decode.

Reply:
Just delete this entire step, decode_module already enforces it.

Reply:
That's primarily a spec for web en/decoding APIs, which the Wasm core does not need and has no reason to depend on. The Unicode spec is the primary source of truth for the few aspects that matter there.

Reply:
Should the name section be validated? I see the specific note,

> If an implementation interprets the contents of a custom section, then errors in that contents, or the placement of the section, must not invalidate the module.

If I just remove this section, I don't see how the names will be validated. OTOH, this text currently doesn't seem sufficient already for validating the names section. I'm drafting text to be more strict about it.

Reply:
Yes, that comment is important, and what you observe is intentional. The name section, like any custom section, is essentially allowed to be totally broken, which of course implies that names occurring in it may be malformed just as well. The JS API must not enforce anything there either. In fact, I see no reason why the name section should even be mentioned in the API spec.

Reply:
OK, adding a note that specifically calls out the name section as not validated.

Reply:
> That's primarily a spec for web en/decoding APIs

That's not correct.

Reply:
The comment has been removed.